[WTF] Unify granule tracking between ArenaAllocator and ImmortalHeap

Achierius opened the pull request:
#### 48eb357acbef689220d358f23821731600eecdd6
<pre>
[WTF] Unify granule tracking between ArenaAllocator and ImmortalHeap
<a href="https://rdar.apple.com/145159083">rdar://145159083</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=287993">https://bugs.webkit.org/show_bug.cgi?id=287993</a>

Reviewed by Yusuke Suzuki.

Previously, granules were essentially implicit to the internal structure
of the SequesteredArenaAllocator. Now, the SequesteredImmortalHeap
expects every allocator to have a ConcurrentDecommitQueue of granules,
which allows the scavenger to target those through the
SequesteredImmortalHeap, allowing us to offload all decommit overhead to
a single central scavenging point.

* Source/WTF/wtf/SequesteredAllocator.h:
(WTF::ArenaGranuleHeader::headerForGranule): Deleted.
* Source/WTF/wtf/SequesteredImmortalHeap.cpp:
(WTF::ConcurrentDecommitQueue::decommit):
* Source/WTF/wtf/SequesteredImmortalHeap.h:
(WTF::GranuleList::GranuleList):
(WTF::GranuleList::operator=):
(WTF::GranuleList::concatenate):
(WTF::GranuleList::push):
(WTF::GranuleList::first):
(WTF::GranuleList::last):
(WTF::GranuleList::popFirst):
(WTF::GranuleList::popLast):
(WTF::GranuleList::empty const):
(WTF::GranuleList::assertBounds const):
(WTF::ConcurrentDecommitQueue::concatenate):
(WTF::ConcurrentDecommitQueue::acquireExclusiveCopyOfGranuleList):
(WTF::SequesteredImmortalHeap::mapGranule):
(WTF::SequesteredImmortalHeap::decommitGranule):
(WTF::SequesteredImmortalHeap::SequesteredImmortalHeap):
(WTF::SequesteredImmortalHeap::mapPages): Deleted.

Canonical link: <a href="https://commits.webkit.org/291240@main">https://commits.webkit.org/291240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcf58fbe6eae51b321ef26473e4255b32bff1e30

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92406 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11947 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97390 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/42913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20408 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/42913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95408 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9294 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83688 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51138 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1316 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/85116 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99416 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91072 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19457 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19707 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79561 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79086 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23634 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14703 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19441 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113720 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19128 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.cpp:
```diff
@@ -29,6 +29,39 @@
+    auto& sih = SequesteredImmortalHeap::instance();
+
+    [[maybe_unused]] size_t decommitPageCount { 0 };
+    [[maybe_unused]] size_t decommitGranuleCount { 0 };
```
Use `UNUSED_VARIABLE`.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -50,6 +51,168 @@
+            return;
+
+        {
+            WTF::Locker lock(m_decommitLock);
```
`WTF::` is not necessary.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -50,6 +51,168 @@
+    }
+
+    GranuleList m_granules { };
+    WTF::Lock m_decommitLock { };
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -50,6 +51,168 @@
+    {
+        GranuleList granules { };
+        {
+            WTF::Locker lock(m_decommitLock);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -50,6 +51,168 @@
+    size_t additionalPageCount;
+};
+
+class GranuleList {
```
How about using existing `DoublyLinkedList`?

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -50,6 +51,168 @@
 
 namespace WTF {
 
+struct GranuleHeader {
```
How about using `DoublyLinkedListNode`?

Achierius replied:
I've done so locally but it seems to have introduced some bugs. At the very least DoublyLinkedList doesn't have move semantics, but even after accounting for that I'm still getting corruptions -- need to investigate further

Achierius committed:
[WTF] Unify granule tracking between ArenaAllocator and ImmortalHeap
rdar://145159083
https://bugs.webkit.org/show_bug.cgi?id=287993

Reviewed by Yusuke Suzuki.

Previously, granules were essentially implicit to the internal structure
of the SequesteredArenaAllocator. Now, the SequesteredImmortalHeap
expects every allocator to have a ConcurrentDecommitQueue of granules,
which allows the scavenger to target those through the
SequesteredImmortalHeap, allowing us to offload all decommit overhead to
a single central scavenging point.

* Source/WTF/wtf/SequesteredAllocator.h:
(WTF::ArenaGranuleHeader::headerForGranule): Deleted.
* Source/WTF/wtf/SequesteredImmortalHeap.cpp:
(WTF::ConcurrentDecommitQueue::decommit):
* Source/WTF/wtf/SequesteredImmortalHeap.h:
(WTF::GranuleList::GranuleList):
(WTF::GranuleList::operator=):
(WTF::GranuleList::concatenate):
(WTF::GranuleList::push):
(WTF::GranuleList::first):
(WTF::GranuleList::last):
(WTF::GranuleList::popFirst):
(WTF::GranuleList::popLast):
(WTF::GranuleList::empty const):
(WTF::GranuleList::assertBounds const):
(WTF::ConcurrentDecommitQueue::concatenate):
(WTF::ConcurrentDecommitQueue::acquireExclusiveCopyOfGranuleList):
(WTF::SequesteredImmortalHeap::mapGranule):
(WTF::SequesteredImmortalHeap::decommitGranule):
(WTF::SequesteredImmortalHeap::SequesteredImmortalHeap):
(WTF::SequesteredImmortalHeap::mapPages): Deleted.

Canonical link: https://commits.webkit.org/291240@main
